pkg/apis/crd/v1beta1/types.go

```diff
@@ -0,0 +1,159 @@
+	// sources.
+	// +optional
+	From []NetworkPolicyPeer `json:"from"`
+}
+
+// EgressRule describes the traffic allowed from the workloads selected by
+// Spec.AppliedTo. Based on the action specified in the rule, traffic is either
+// allowed or denied which exactly match the specified ports and protocol.
+type EgressRule struct {
+	// Action specifies the action to be applied on the rule. Defaults to
+	// ALLOW action.
+	// +optional
+	Action *RuleAction `json:"action"`
+	// Set of port and protocol allowed/denied by the rule. If this field is unset
+	// or empty, this rule matches all ports.
+	// +optional
+	Ports []NetworkPolicyPort `json:"ports"`
+	// Rule is matched if traffic is intended for workloads selected by
+	// this field. If this field is empty or missing, this rule matches all
+	// destinations.
+	// +optional
+	To []NetworkPolicyPeer `json:"to"`
+}
+
+// NetworkPolicyPeer describes the grouping selector of workloads.
+type NetworkPolicyPeer struct {
+	// IPBlock describes the IPAddresses/IPBlocks that is matched in to/from.
+	// IPBlock cannot be set as part of the AppliedTo field
+	// Cannot be set with any other selector.
+	// +optional
+	IPBlock *IPBlock `json:"ipBlock,omitempty"`
+	// Select Pods from all Namespaces as workloads in AppliedTo/To/From
+	// fields. If set with NamespaceSelector, Pods are matched from Namespaces
+	// matched by the NamespaceSelector.
+	// Cannot be set with any other selector except NamespaceSelector.
+	// +optional
+	PodSelector *metav1.LabelSelector `json:"podSelector,omitempty"`
+	// Select all Pods from Namespaces matched by this selector, as
+	// workloads in AppliedTo/To/From fields. If set with PodSelector,
+	// Pods are matched from Namespaces matched by the NamespaceSelector.
+	// Cannot be set with any other selector except PodSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// IPBlock describes a particular CIDR (Ex. "192.168.1.1/24") that is allowed
+// or denied to/from the workloads matched by a Spec.AppliedTo.
+type IPBlock struct {
+	// CIDR is a string representing the IP Block
+	// Valid examples are "192.168.1.1/24".
+	CIDR string `json:"cidr"`
+}
+
+// NetworkPolicyPort describes the port and protocol to match in a rule.
+type NetworkPolicyPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match.
+	// If not specified, this field defaults to TCP.
+	// +optional
+	Protocol *v1.Protocol `json:"protocol"`
+	// The port on the given protocol. This can either be a numerical
+	// or named port on a Pod. If this field is not provided, this
+	// matches all port names and numbers.
+	// TODO: extend it to include Port Range.
+	// +optional
+	Port *intstr.IntOrString `json:"port"`
+}
+
+// RuleAction describes the action to be applied on traffic matching a rule.
+type RuleAction string
+
+// ToUpper returns a RuleAction type in uppercase.
+func (a *RuleAction) ToUpper() RuleAction {
```

Comment:
what's this used for?

Reply:
for comparison, user may input Allow/ALLOW/allow...

Reply:
I didn't see it's used, actually I'm a little confused, the method has a receiver of RuleAction, then it's already either RuleActionAllow or RuleActionDrop before calling this `ToUpper`? why it needs to be converted once? 

BTW, I haven't saw any methods defined in types.go in K8s, maybe let's follow the convention and define functions in other files of this package like https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/taint.go if they are really needed?

Reply:
its used in the follow up PR.. removed for now.. we can discuss whether needed in the controller PR